header/p2p/session.go

```diff
@@ -0,0 +1,228 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/header"
+	p2p_pb "github.com/celestiaorg/celestia-node/header/p2p/pb"
+)
+
+// TODO(@vgonkivs): make it configurable
+var (
+	// headersPerPeer is a maximum amount of headers that will be requested per peer.
+	headersPerPeer uint64 = 64
+	// sessionDuration is an amount of time after which session will be closed.
+	sessionDuration = time.Minute * 2
+)
+
+// session aims to divide a range of headers
+// into several smaller requests among different peers.
+type session struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	host       host.Host
+	protocolID protocol.ID
+	// peerTracker contains discovered peers with records that describes their activity.
+	queue *peerQueue
+
+	reqCh chan *p2p_pb.ExtendedHeaderRequest
+	errCh chan error
+}
+
+func newSession(ctx context.Context, h host.Host, peerTracker []*peerStat, protocolID protocol.ID) *session {
+	ctx, cancel := context.WithTimeout(ctx, sessionDuration)
```

Comment:
I insist on consistency here. As we do with `Head` we should block the operation forever or until the requests are fulfilled, and only allow users to set timeout as per their wll

Reply:
ok. Then, to be consistent with `Head` method, I suggest to block forever... In case, when we will need such mechanism, then we will implement it for both `Head` and `GetRangeByHeight` in one PR.

@renaynay wdyt?

Reply:
Yeah on client side, it's fine.